src/libtiled/mapreader.cpp

```diff
@@ -921,36 +964,72 @@ MapObject *MapReaderPrivate::readObject()
     const QPointF pos(x, y);
     const QSizeF size(width, height);
 
-    MapObject *object = new MapObject(name, type, pos, size);
+    MapObject *object = new MapObject;
+
+    if (tid) { // This object is a template instance
+        ObjectTemplate *objectTemplate = templateForTid(tid);
+        if (objectTemplate) {
+            object->setTemplateRef({objectTemplate->templateGroup(), objectTemplate->id()});
```

Comment:
I think with how I implemented the tids, keeping them valid will be more complicated because If a tid wasn't mapped to any object, the templateRef won't be valid, and the tid will be lost forever.

I think the cause of the problem is, unlike tiles in a tileset, every template in the group has it's own id, which is part of the templateRef, which is used when the tid is created, so maybe we need a new way of how the templateRef works.